src/RequestHandler/environmentvariablehelpers.h

```diff
@@ -0,0 +1,403 @@
+        _In_ BOOL                           fWebsocketsEnabled,
+        _In_ APP_HOSTING_MODEL              hostingModel,
+        _Out_ ENVIRONMENT_VAR_HASH**        ppEnvironmentVarTable
+    )
+    {
+        HRESULT hr = S_OK;
+        BOOL    fFound = FALSE;
+        DWORD   dwResult, dwError;
+        STRU    strIisAuthEnvValue;
+        STACK_STRU(strStartupAssemblyEnv, 1024);
+        STACK_STRU(strIISWebsocketEnvValue, 40);
+        ENVIRONMENT_VAR_ENTRY* pHostingEntry = NULL;
+        ENVIRONMENT_VAR_ENTRY* pIISAuthEntry = NULL;
+        ENVIRONMENT_VAR_ENTRY* pIISWebsocketEntry = NULL;
+        ENVIRONMENT_VAR_HASH* pEnvironmentVarTable = NULL;
+
+        pEnvironmentVarTable = new ENVIRONMENT_VAR_HASH();
+        if (pEnvironmentVarTable == NULL)
+        {
+            hr = E_OUTOFMEMORY;
+            goto Finished;
+        }
+
+        //
+        // few environment variables expected, small bucket size for hash table
+        //
+        if (FAILED(hr = pEnvironmentVarTable->Initialize(37 /*prime*/)))
+        {
+            goto Finished;
+        }
+
+        // copy the envirable hash table (from configuration) to a temp one as we may need to remove elements 
+        pInEnvironmentVarTable->Apply(ENVIRONMENT_VAR_HELPERS::CopyToTable, pEnvironmentVarTable);
+        if (pEnvironmentVarTable->Count() != pInEnvironmentVarTable->Count())
+        {
+            // hash table copy failed
+            hr = E_UNEXPECTED;
+            goto Finished;
+        }
+
+        pEnvironmentVarTable->FindKey((PWSTR)ASPNETCORE_IIS_AUTH_ENV_STR, &pIISAuthEntry);
+        if (pIISAuthEntry != NULL)
+        {
+            // user defined ASPNETCORE_IIS_HTTPAUTH in configuration, wipe it off
+            pIISAuthEntry->Dereference();
+            pEnvironmentVarTable->DeleteKey((PWSTR)ASPNETCORE_IIS_AUTH_ENV_STR);
+        }
+
+        if (fWindowsAuthEnabled)
+        {
+            strIisAuthEnvValue.Copy(ASPNETCORE_IIS_AUTH_WINDOWS);
+        }
+
+        if (fBasicAuthEnabled)
+        {
+            strIisAuthEnvValue.Append(ASPNETCORE_IIS_AUTH_BASIC);
+        }
+
+        if (fAnonymousAuthEnabled)
+        {
+            strIisAuthEnvValue.Append(ASPNETCORE_IIS_AUTH_ANONYMOUS);
+        }
+
+        if (strIisAuthEnvValue.IsEmpty())
+        {
+            strIisAuthEnvValue.Copy(ASPNETCORE_IIS_AUTH_NONE);
+        }
+
+        pIISAuthEntry = new ENVIRONMENT_VAR_ENTRY();
+        if (pIISAuthEntry == NULL)
+        {
+            hr = E_OUTOFMEMORY;
+            goto Finished;
+        }
+        if (FAILED(hr = pIISAuthEntry->Initialize(ASPNETCORE_IIS_AUTH_ENV_STR, strIisAuthEnvValue.QueryStr())) ||
+            FAILED(hr = pEnvironmentVarTable->InsertRecord(pIISAuthEntry)))
+        {
+            goto Finished;
+        }
+
+        // We only need to set the WEBSOCKET_SUPPORTED environment variable for out of process
+        if (hostingModel == HOSTING_OUT_PROCESS)
```

Comment:
The only change made to this file from before is here @pan-wang

Reply:
It's a little odd to pass in hostingModel only for this condition. It seems like fWebsocketsEnabled should be a tri-state field: enabled, disabled, no-op.

Reply:
yeah, I'll clean that up.